Add Blueprint and integration test app for CockroachDB (#1585)
kanisterio/kanister: examples/cockroachdb/README.md

# CockroachDB

[CockroachDB](https://www.cockroachlabs.com/) is one of the database servers based on SQL.

## Introduction

This chart bootstraps a three node Cockroachdb deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

## Prerequisites

- Kubernetes 1.16+ with Beta APIs enabled
- PV provisioner support in the underlying infrastructure
- Kanister controller version 0.80.0 installed in your cluster, let's assume in Namespace `kanister`
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)

## Installing the Chart

To install the CockroachDB database using the `cockroachdb` chart with the release name `cockroachdb-release`:

```bash
# Add cockroachdb in your local chart repository
$ helm repo add cockroachdb https://charts.cockroachdb.com/

# Update your local chart repository
$ helm repo update

# Install the cockroachdb in cockroachdb namespace
$ helm install cockroachdb-release --namespace cockroachdb --create-namespace cockroachdb/cockroachdb

# Install the secure cockroachdb client

> $ curl -OOOOOOOOO https://raw.githubusercontent.com/cockroachdb/helm-charts/master/examples/client-secure.yaml

Reviewer: Instead of referring to the file from the master, we should refer it from the latest release.

# Edit Fields in client-secure.yaml
spec.serviceAccountName: cockroachdb-release

> spec.image: cockroachdb/cockroach:<VERSION>

Reviewer: What are the correct possible versions one can use?
Can we provide a sample yaml manifest?

spec.volumes[0].project.sources[0].secret.name: cockroachdb-release-client-secret

$ kubectl create -f ./client-secure.yaml -n cockroachdb

```
> **Note** : 
cockroachdb/cockroachdb version in spec.image as mentioned in following chart [values.yaml](https://github.com/cockroachdb/helm-charts/blob/master/cockroachdb/values.yaml)

The command deploys a CockroachDB instance in the `cockroachdb` namespace.

> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

## Integrating with Kanister

If you have deployed CockroachDB application with name other than `cockroachdb-release` and namespace other than `cockroachdb`, you need to modify the commands(backup, restore and delete) used below to use the correct release name and namespace

### Create Profile

> Create Profile if not created already and Edit the values accordingly

Reviewer 2: ```suggestion
Create Profile if not created already and set the values accordingly
```

```bash
$ kanctl create profile s3compliant --access-key <aws-access-key-id> \
	--secret-key <aws-secret-key> \
	--bucket <s3-bucket-name> --region <region-name> \
	--namespace kanister
```

You can read more about the Profile custom Kanister resource [here](https://docs.kanister.io/architecture.html?highlight=profile#profiles).

**NOTE:**

The above command will configure a location where artifacts resulting from Kanister
data operations such as backup should go. This is stored as a `profiles.cr.kanister.io`
*CustomResource (CR)* which is then referenced in Kanister ActionSets. Every ActionSet
requires a Profile reference to complete the action. This CR (`profiles.cr.kanister.io`)
can be shared between Kanister-enabled application instances.

### Create Blueprint

Create Blueprint in the same namespace as the Kanister controller

```bash

> $ kubectl create -f ./cockroachdb-blueprint.yaml

Reviewer 2: ```suggestion
$ kubectl create -f ./cockroachdb-blueprint.yaml -n kansiter
```

# CockroachDB

[CockroachDB](https://www.cockroachlabs.com/) is one of the database servers based on SQL.

## Introduction

This chart bootstraps a three node CockroachDB deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

## Prerequisites

- Kubernetes 1.16+ with Beta APIs enabled
- PV provisioner support in the underlying infrastructure
- Kanister controller version 0.80.0 installed in your cluster, let's assume in Namespace `kanister`
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)

## Installing the Chart

The commands mentioned below could be run to install the CockroachDB database using the `cockroachdb` chart with the release name `cockroachdb-release`:

```bash
# Add cockroachdb in your local chart repository
$ helm repo add cockroachdb https://charts.cockroachdb.com/

# Update your local chart repository
$ helm repo update

# Install the cockroachdb in cockroachdb namespace
$ helm install cockroachdb-release --namespace cockroachdb --create-namespace cockroachdb/cockroachdb

# Install the secure cockroachdb client

> # Download latest secure client manifest 

Author: we're using cockroachdb secure client, so that user could interact with the cockroachdb cluster. Initially user would have to do some activity like insert data into tables etc. so for that user have to create a secure client pod, from the blueprint perspective, we're dynamically creating the client pods using KubeTask

$ curl -O https://raw.githubusercontent.com/cockroachdb/helm-charts/master/examples/client-secure.yaml

# Edit Fields in client-secure.yaml. A sample manifest file is being included for reference.
# Value of spec.serviceAccountName should be same as your release name
spec.serviceAccountName: cockroachdb-release
# Find the latest tag of image cockroachdb/cockroach from https://github.com/cockroachdb/helm-charts/blob/master/cockroachdb/values.yaml
spec.image: cockroachdb/cockroach:<VERSION>
# Value of spec.volumes[0].project.sources[0].secret.name should be <release-name>-client-secret
spec.volumes[0].project.sources[0].secret.name: cockroachdb-release-client-secret

$ kubectl create -f ./client-secure.yaml -n cockroachdb
```

> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

## Integrating with Kanister

> If you have deployed CockroachDB application with name other than `cockroachdb-release` and namespace other than `cockroachdb`, you need to modify the commands (backup, restore and delete) used below to use the correct release name and namespace

Reviewer 3: ```suggestion
If you have deployed CockroachDB application with name other than `cockroachdb-release` and namespace other than `cockroachdb`, you need to modify the commands (backup, restore and delete) used below to use the correct release name and namespace.

```

Author: Done !

  bank          | root  | NULL           | {}      | NULL
  defaultdb     | root  | NULL           | {}      | NULL
  postgres      | root  | NULL           | {}      | NULL
  system        | node  | NULL           | {}      | NULL
(4 rows)

> DROP DATABASE bank CASCADE;
Query OK, 1 row affected (0.03 sec)

> SHOW DATABASES;
  database_name | owner | primary_region | regions | survival_goal
----------------+-------+----------------+---------+----------------
  defaultdb     | root  | NULL           | {}      | NULL
  postgres      | root  | NULL           | {}      | NULL
  system        | node  | NULL           | {}      | NULL
(3 rows)

```

### Restore the Application

To restore the missing data, you should use the backup that you created before. An easy way to do this is to leverage `kanctl`, a command-line tool that helps create ActionSets that depend on other ActionSets:
> **Note** : CockroachDB uses [Garbage Collection](https://www.cockroachlabs.com/docs/stable/architecture/storage-layer.html#garbage-collection), So in order to perform Full Cluster Restore on same cluster, one must wait out the Default Garbage Collection Run After Time i.e. 90,000 Seconds or 25 Hours.

> > On the other hand the following step is Not Required if one is restoring data on a new fresh cluster.

Reviewer 3: ```suggestion
> On the other hand the following step is Not Required if snapshot is being restored on a new DB cluster.
```

> > In Order to reduce the Garbage Collection Time, run the following command, Wait for a minute and let cockroachdb to automatic collect garbage, and then restore

Reviewer 3: ```suggestion
> In Order to reduce the Garbage Collection time, run the following command, wait for a minute and let cockroachdb to automatic garbage collection. And then use kanctl to create restore action.
```

Author: Done !
